api/v1alpha1/externaldns_types.go

```diff
@@ -0,0 +1,146 @@
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems:=1
+	// +kubebuilder:validation:MaxItems:=7
+	// +kubebuilder:validation:items:UniqueItems:=true
+	// +kubebuilder:validation:items:MaxLength:=172
+	// +kubebuilder:validation:items:MinLength:=103
+	// +kubebuilder:validation:items:MaxProperties:=1
+	// +kubebuilder:validation:XValidation:rule="self.all(item, item.split('/')[2] == self[0].split('/')[2])",message="all items must have the same subscription ID"
+	// +kubebuilder:validation:XValidation:rule="self.all(item, item.split('/')[4] == self[0].split('/')[4])",message="all items must have the same resource group"
+	// +kubebuilder:validation:XValidation:rule="self.all(item, item.split('/')[7] == self[0].split('/')[7])",message="all items must be of the same resource type"
+	// +listType:=set
+	DNSZoneResourceIDs []string `json:"dnsZoneResourceIDs"`
+
+	// ResourceTypes is a list of Kubernetes resource types that the ExternalDNS controller should manage. The supported resource types are 'ingress' and 'gateway'.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:MinItems:=1
+	// +kubebuilder:validation:MaxItems:=2
+	// +kubebuilder:validation:XValidation:rule="self.all(item, item.matches('(?i)(gateway|ingress)'))",message="all items must be either 'gateway' or 'ingress'"
+	// +listType:=set
+	ResourceTypes []string `json:"resourceTypes"`
+
+	// Identity contains information about the identity that ExternalDNS will use to interface with Azure resources.
+	// +kubebuilder:validation:Required
+	Identity ExternalDNSIdentity `json:"identity"`
+
+	// Filters contains optional filters that the ExternalDNS controller should use to determine which resources to manage.
+	// +optional
+	Filters ExternalDNSFilters `json:"filters,omitempty"`
+}
+
+// ExternalDNSIdentity contains information about the identity that ExternalDNS will use to interface with Azure resources.
+type ExternalDNSIdentity struct {
+	// +kubebuilder:validation:MinLength=1
+	// +kubebuilder:validation:MaxLength=253
+	// +kubebuilder:validation:Pattern=`^[a-z0-9][-a-z0-9\.]*[a-z0-9]$`
+	// +kubebuilder:validation:Required
+	// ServiceAccount is the name of the Kubernetes ServiceAccount that ExternalDNS will use to interface with Azure resources. It must be in the same namespace as the ExternalDNS.
+	ServiceAccount string `json:"serviceAccount"`
+}
+
+type ExternalDNSFilters struct {
+	// GatewayLabels contains key-value pairs that the ExternalDNS controller will use to filter the Gateways that it manages.
+	// +optional
+	GatewayLabels map[string]string `json:"gatewayLabels,omitempty"`
+
+	// RouteAndIngressLabels contains key-value pairs that the ExternalDNS controller will use to filter the HTTPRoutes and Ingresses that it manages.
+	// +optional
+	RouteAndIngressLabels map[string]string `json:"routeAndIngressLabels,omitempty"`
```

Comment:
why are route and ingress ones combined but gateway ones aren't?

Reply:
externaldns has a separate argument for gateway but uses the same argument for both ingresses and routes - 

gateway specific filter is documented [here](https://kubernetes-sigs.github.io/external-dns/v0.14.1/sources/gateway/#matching-gateways)

meanwhile, there's no mention of an ingress-label-filter on the [ingress page](https://kubernetes-sigs.github.io/external-dns/v0.14.1/sources/ingress/), and there's no page for routes on the sources page, but the [gateway tutorial](https://kubernetes-sigs.github.io/external-dns/v0.14.1/tutorials/gateway-api/#manifest-with-rbac) just comments that the "--label-filter" argument would apply to routes